Importación de librerías

In [1]:
import pandas as pd
import numpy as np


In [6]:
df_2012 = pd.read_excel('../../data/output/customer-details.xlsx', sheet_name="2012")
df_2012.sample(2)


Unnamed: 0  Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
7879        7879   28320        0         2  2012-06-11                 19   
5778        5778   33789        0         0  2012-03-21                 12   

                                        ID  
7879  f8d3287d-9e79-4171-81ec-6411a1fee186  
5778  e6c38ae4-0037-4671-9dba-1b047baf649a

In [12]:
df_2013 =pd.read_excel('../../data/output/customer-details.xlsx', sheet_name="2013")
df_2013.sample(2)
df_2013.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8965 entries, 0 to 8964
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Unnamed: 0         8965 non-null   int64         
 1   Income             8965 non-null   int64         
 2   Kidhome            8965 non-null   int64         
 3   Teenhome           8965 non-null   int64         
 4   Dt_Customer        8965 non-null   datetime64[ns]
 5   NumWebVisitsMonth  8965 non-null   int64         
 6   ID                 8965 non-null   object        
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 490.4+ KB


In [11]:
df_2014 = pd.read_excel('../../data/output/customer-details.xlsx', sheet_name="2014")
df_2014.sample(2)
df_2014.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14090 entries, 0 to 14089
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Unnamed: 0         14090 non-null  int64         
 1   Income             14090 non-null  int64         
 2   Kidhome            14090 non-null  int64         
 3   Teenhome           14090 non-null  int64         
 4   Dt_Customer        14090 non-null  datetime64[ns]
 5   NumWebVisitsMonth  14090 non-null  int64         
 6   ID                 14090 non-null  object        
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 770.7+ KB


In [13]:
df_2012.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20115 entries, 0 to 20114
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Unnamed: 0         20115 non-null  int64         
 1   Income             20115 non-null  int64         
 2   Kidhome            20115 non-null  int64         
 3   Teenhome           20115 non-null  int64         
 4   Dt_Customer        20115 non-null  datetime64[ns]
 5   NumWebVisitsMonth  20115 non-null  int64         
 6   ID                 20115 non-null  object        
dtypes: datetime64[ns](1), int64(5), object(1)
memory usage: 1.1+ MB


Revisamos la informacion de cada DF para verificar que las columnas son las mismas, el tipo de dato de cada columna y si hay o no valores nulos (en este caso no los hay)

Unificaremos las distintas hojas del archivo Excel en un único DF, cada hoja se corresponde con años consecutivos, de 2012 a 2014. Mediante la función pd.concat(), ya que todas presentan la misma estructura de variables.

In [14]:
df= pd.concat([df_2012, df_2013, df_2014], ignore_index=True)


In [15]:
df.sample(2)

Unnamed: 0  Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
20640         525  121809        0         1  2013-06-13                 28   
22800        2685   43392        0         1  2013-04-04                  7   

                                         ID  
20640  1a29bbfd-46d6-473a-90d9-2bc9826a0be3  
22800  3b24f53d-726d-4b90-918c-d890482d8b39

Eliminamos la columna "Unnamed: 0" dado que no nos aporta valos al análisis

In [21]:
df=df.drop(columns="Unnamed: 0")

In [22]:
df.describe().T

count                           mean  \
Income             43170.0                   93227.389622   
Kidhome            43170.0                       1.004749   
Teenhome           43170.0                       0.998633   
Dt_Customer          43170  2013-05-11 22:09:47.407922176   
NumWebVisitsMonth  43170.0                       16.59129   

                                   min                  25%  \
Income                          5841.0              49608.0   
Kidhome                            0.0                  0.0   
Teenhome                           0.0                  0.0   
Dt_Customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
NumWebVisitsMonth                  1.0                  9.0   

                                   50%                  75%  \
Income                         93009.5             136740.5   
Kidhome                            1.0                  2.0   
Teenhome                           1.0                  2.0   
Dt_Customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
NumWebVisitsMonth                 17.0                 25.0   

                                   max           std  
Income                        180802.0  50498.181989  
Kidhome                            2.0      0.815996  
Teenhome                           2.0      0.816008  
Dt_Customer        2014-12-31 00:00:00           NaN  
NumWebVisitsMonth                 32.0      9.241769

In [31]:
for col in df.select_dtypes(include="O").columns:
    df[col] = df[col].str.lower()

In [32]:
df.sample()

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
4184  170474        1         1  2012-08-13                 11   

                                        ID   año  
4184  cbdb95d2-05b0-4272-b7b8-76c1496d92e6  2012

Extraemos el año en una nueva columna, con el fin de facilitar los análisis temporales y las comparaciones interanuales.

In [23]:
df["año"] = df["Dt_Customer"].dt.year

In [24]:
df.sample()

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
25096   25978        0         1  2013-08-16                 24   

                                         ID   año  
25096  9d93fc3f-bc5f-4666-a5c1-233612de6f10  2013

In [29]:
df["Kidhome"].value_counts()


Kidhome
2    14475
1    14425
0    14270
Name: count, dtype: int64

In [30]:
df["Teenhome"].value_counts()

Teenhome
1    14425
0    14402
2    14343
Name: count, dtype: int64

Se normalizan los nombres de las variables a minúscula para garantizar consistencia con el DF "bank-additional", respetando la columna "ID" clave para la identificación para la posterior unión.

In [42]:

df.columns= [col.lower() if col != "ID" else col
for col in df.columns]

In [43]:
df.sample(2)

income  kidhome  teenhome dt_customer  numwebvisitsmonth  \
19014  113181        0         1  2012-09-19                  7   
38864  101362        0         0  2014-04-26                  1   

                                         ID   año  
19014  f40b2ad4-02b8-414b-ac1d-2c1f163e6d56  2012  
38864  13618323-f6b6-4c50-904a-c19032ab9f8f  2014

In [44]:
df.duplicated().sum()


np.int64(0)

No hay duplicados en este DataFrame

In [46]:
df.dtypes

income                        int64
kidhome                       int64
teenhome                      int64
dt_customer          datetime64[ns]
numwebvisitsmonth             int64
ID                           object
año                           int32
dtype: object

In [47]:
df.groupby("kidhome")["income"].mean()

kidhome
0    93034.309250
1    93441.314939
2    93204.549154
Name: income, dtype: float64

In [48]:
df.groupby("teenhome")["income"].mean()

teenhome
0    92824.058672
1    93237.215529
2    93622.497595
Name: income, dtype: float64

En un primer análisis sencillo, vemos que a medida que sube el numero de hijos en casa, sube el nivel medio de ingresos. Ahora queremos comprobar si las personas que más utilizan la aplicacion del banco son aquellas que más dinero tienen.

In [53]:
df.groupby("numwebvisitsmonth")["income"].mean().sort_index(ascending=False)

numwebvisitsmonth
32    94245.931352
31    91913.021614
30    93770.285500
29    93323.989177
28    93834.724467
27    95947.810117
26    93147.400000
25    93091.055514
24    91865.961049
23    92860.182482
22    93824.069578
21    92369.126893
20    93076.020543
19    92207.916168
18    94925.087938
17    95373.124812
16    93877.787645
15    93477.275938
14    91301.457283
13    91761.088580
12    92548.028633
11    93477.335380
10    91845.619677
9     90648.015766
8     93780.650038
7     92311.000000
6     94142.941132
5     95257.752252
4     94620.294420
3     93687.944759
2     93409.198431
1     91539.807605
Name: income, dtype: float64

Vemos que la variable "veces que entra en la web" no esta relacionada con el nivel salarial que los clientes tienen, ya que vemos que para el valor "4" la media salarial es de  94620.29€ y para el valor "32" la media salarial es de 94245.93€

Dado que ya hemos revisado de forma indidual los datos de cada DF antes de la unión y que ya hemos hecho los cambios oportunos en el DF unido, procedemos a guardar estos datos en un nuevo archivo csv para proceder a la union con el csv ya trabajado "bank-additional.csv"

In [58]:
df.to_csv("../../data/output/customer-details-unido.csv", index=False)